azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -609,11 +609,15 @@ private void doRun() {
         "Finishing job " + this.jobId + getNodeRetryLog() + " at " + this.node.getEndTime()
             + " with status " + this.node.getStatus());
 
-    fireEvent(Event.create(this, EventType.JOB_FINISHED,
-        new EventData(finalStatus, this.node.getNestedId())), false);
-    finalizeLogFile(this.node.getAttempt());
-    finalizeAttachmentFile();
-    writeStatus();
+    try {
+      finalizeLogFile(this.node.getAttempt());
+      finalizeAttachmentFile();
+      writeStatus();
+    } finally {
+      // note that FlowRunner thread does node.attempt++ when it receives the JOB_FINISHED event
```

Comment:
Do you mean there still exist possibilities that writing log to db would fail due to race condition?

Reply:
I'm not aware of any specific case at the moment, but it's really important to guarantee that JOB_FINISHED event is fired. Otherwise the flow execution remains stuck in RUNNING state without ever finishing. For example, if writing the logs fails because of any random reason (for example DB connectivity) it's not as severe as having the flow get stuck.

Reply:
Makes sense. Thanks for your contribution!